#Bereinigung des zusammengefügten Datensatzes

Nach dem wir alle Datensätze gejoint haben und verschiedene Joins ausprobiert bereinigen wir den über den Outer Join zusammengefügten Datensatz. Dieser Join hat am besten funktioniert, die anderen hatten nur noch wenige Zeile oder viele fehlende Werte.

In [1]:
#Daten laden

import pandas as pd

df = pd.read_csv('Join_5_movie_meta_data.csv')

<ipython-input-1-a09236d8848b>:5: DtypeWarning: Columns (0,1,2,3,4,5,7,8,9,14,16,17,18,19,21,22,23,24,25,26,27,28,29,31,32,33,34,35,36,37) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Join_5_movie_meta_data.csv')


In [2]:
df

id_x                name                  character   role  \
0      tm82169  Sylvester Stallone      Robert "Rocky" Balboa  ACTOR   
1      tm82169         Talia Shire  Adrianna "Adrian" Pennino  ACTOR   
2      tm82169          Burt Young             Paulie Pennino  ACTOR   
3      tm82169       Carl Weathers               Apollo Creed  ACTOR   
4      tm82169    Burgess Meredith            Mickey Goldmill  ACTOR   
...        ...                 ...                        ...    ...   
98821      NaN                 NaN                        NaN    NaN   
98822      NaN                 NaN                        NaN    NaN   
98823      NaN                 NaN                        NaN    NaN   
98824      NaN                 NaN                        NaN    NaN   
98825      NaN                 NaN                        NaN    NaN   

                            title description  runtime genres_x  \
0                             NaN         NaN      NaN      NaN   
1                             NaN         NaN      NaN      NaN   
2                             NaN         NaN      NaN      NaN   
3                             NaN         NaN      NaN      NaN   
4                             NaN         NaN      NaN      NaN   
...                           ...         ...      ...      ...   
98821     While You Were Sleeping         NaN      NaN      NaN   
98822                  Wild Tales         NaN      NaN      NaN   
98823                  Wind River         NaN      NaN      NaN   
98824          Wrongfully Accused         NaN      NaN      NaN   
98825  You Were Never Really Here         NaN      NaN      NaN   

      production_countries_x imdb_id_x  ...  \
0                        NaN       NaN  ...   
1                        NaN       NaN  ...   
2                        NaN       NaN  ...   
3                        NaN       NaN  ...   
4                        NaN       NaN  ...   
...                      ...       ...  ...   
98821                    NaN       NaN  ...   
98822                    NaN       NaN  ...   
98823                    NaN       NaN  ...   
98824                    NaN       NaN  ...   
98825                    NaN       NaN  ...   

                                        country  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
...                                         ...   
98821                             United States   
98822  Argentina, Spain, France, United Kingdom   
98823     United Kingdom, Canada, United States   
98824                    Germany, United States   
98825     United Kingdom, United States, France   

                                genres  imdb_id_y  \
0                                  NaN        NaN   
1                                  NaN        NaN   
2                                  NaN        NaN   
3                                  NaN        NaN   
4                                  NaN        NaN   
...                                ...        ...   
98821           Comedy, Drama, Romance   114924.0   
98822          Comedy, Drama, Thriller  3011894.0   
98823  Crime, Drama, Mystery, Thriller  5362988.0   
98824         Action, Comedy, Thriller   120901.0   
98825  Crime, Drama, Mystery, Thriller  5742374.0   

                                                  awards  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
98821  Golden Globe 1996, American Comedy Award 1996,...   
98822  Oscar 2015, BAFTA Film Award 2016, Awa

In [3]:
#Prüfen nach fehlenden Werten

df.isna().sum()

id_x                                                 16407
name                                                 17471
character                                            28294
role                                                 17471
title                                                17545
description                                          25412
runtime                                              20579
genres_x                                             33951
production_countries_x                               33951
imdb_id_x                                            36938
imdb_score                                           34570
imdb_votes                                           34643
tmdb_score                                           30417
budget_x                                             93615
genres_y                                             93615
id_y                                                 93615
original_language                                    936

Als erstes werden wir einige Spalten, die wir nicht brauchen löschen

In [4]:
df = df.drop(['id_x', 'imdb_id_x', 'Unnamed: 0', 'imdb_id_y'], 1)

<ipython-input-4-b97db58be5d7>:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop(['id_x', 'imdb_id_x', 'Unnamed: 0', 'imdb_id_y'], 1)


Es gibt viele fehlende Werte in dem Datensatz, die durch das zusammenfügen der Datensätze entstanden sind.

In [5]:
df.isna().sum()

name                                                 17471
character                                            28294
role                                                 17471
title                                                17545
description                                          25412
runtime                                              20579
genres_x                                             33951
production_countries_x                               33951
imdb_score                                           34570
imdb_votes                                           34643
tmdb_score                                           30417
budget_x                                             93615
genres_y                                             93615
id_y                                                 93615
original_language                                    93615
production_companies                                 93615
production_countries_y                               936

In [6]:
#Zunächst löschen wir die fehlenden Werte in der title Spalte und der imdb und tmdb score Spalten, da diese für unsere Modell am wichitgsten sind

df = df.dropna(subset = ['title', 'imdb_score', 'tmdb_score'], axis=0)

In [7]:
df.isna().sum()

name                                                   849
character                                             8076
role                                                   849
title                                                    0
description                                             22
runtime                                                  0
genres_x                                                 0
production_countries_x                                   0
imdb_score                                               0
imdb_votes                                              71
tmdb_score                                               0
budget_x                                             60047
genres_y                                             60047
id_y                                                 60047
original_language                                    60047
production_companies                                 60047
production_countries_y                               600

In [8]:
df

name                                          character  \
17545  Graham Chapman  King Arthur / Voice of God / Middle Head / Hic...   
17546     John Cleese  Second Swallow-Savvy Guard / The Black Knight ...   
17547       Eric Idle  Dead Collector / Peasant 1 / Sir Robin the Not...   
17548   Terry Gilliam  Patsy / Green Knight / Old Man from Scene 24 (...   
17549     Terry Jones  Dennis's Mother / Sir Bedevere / Left Head / P...   
...               ...                                                ...   
82683             NaN                                                NaN   
82687             NaN                                                NaN   
82689             NaN                                                NaN   
82691             NaN                                                NaN   
82695             NaN                                                NaN   

        role                               title  \
17545  ACTOR     Monty Python and the Holy Grail   
17546  ACTOR     Monty Python and the Holy Grail   
17547  ACTOR     Monty Python and the Holy Grail   
17548  ACTOR     Monty Python and the Holy Grail   
17549  ACTOR     Monty Python and the Holy Grail   
...      ...                                 ...   
82683    NaN                   The Wedding Coach   
82687    NaN                        Just in Time   
82689    NaN                       Pitta Kathalu   
82691    NaN   Super Monsters: Once Upon a Rhyme   
82695    NaN  Mighty Little Bheem: Kite Festival   

                                             description  runtime  \
17545  King Arthur, accompanied by his squire, recrui...     91.0   
17546  King Arthur, accompanied by his squire, recrui...     91.0   
17547  King Arthur, accompanied by his squire, recrui...     91.0   
17548  King Arthur, accompanied by his squire, recrui...     91.0   
17549  King Arthur, accompanied by his squire, recrui...     91.0   
...                                                  ...      ...   
82683  Former bride and forever comedian Jamie Lee sh...     33.0   
82687  A meticulous clerk for a bookstore that has ju...     91.0   
82689  Four different women, four journeys of love an...     37.0   
82691  The Super Monsters rethink exemplary fantasies...     25.0   
82695  With winter behind them, Bheem and his townspe...      7.0   

                                genres_x production_countries_x  imdb_score  \
17545    ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17546    ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17547    ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17548    ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17549    ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
...                                  ...                    ...         ...   
82683              ['comedy', 'reality']                 ['US']         5.0   
82687      ['comedy', 'drama', 'family']                 ['KE']         6.9   
82689               ['drama', 'romance']                 ['IN']         5.1   
82691            ['animation', 'family']                     []         5.6   
82695  ['family', 'animation', 'comedy']                     []         7.8   

       imdb_votes  ...  cast  country genres  awards producers budget_y  \
17545    534486.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
17546    534486.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
17547    534486.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
17548    534486.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
17549    534486.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
...           ...  ...   ...      ...    ...     ...       ...      ...   
82683        87.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
82687        46.0  ...   NaN      NaN    NaN     NaN       NaN      NaN   
82689       738.0  ...   N

In [9]:
#Um auf einen Datensatz ohne fehlende Werte zu kommen müssen wir noch einige Spalten löschen, in denen es nur wenige Werte gibt

df = df.drop(['budget_x', 'genres_y', 'id_y', 'original_language', 'production_companies', 'production_countries_y', 'tagline', 'awards', 'producers', 'budget_y', 'production companies', 'plot', 'plot outline', 'tageline'], 1)

<ipython-input-9-63752f528401>:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop(['budget_x', 'genres_y', 'id_y', 'original_language', 'production_companies', 'production_countries_y', 'tagline', 'awards', 'producers', 'budget_y', 'production companies', 'plot', 'plot outline', 'tageline'], 1)


In [10]:
df.isna().sum()

name                                                   849
character                                             8076
role                                                   849
title                                                    0
description                                             22
runtime                                                  0
genres_x                                                 0
production_countries_x                                   0
imdb_score                                               0
imdb_votes                                              71
tmdb_score                                               0
Academy Awards adapted screenplay                    59547
Academy Awards original screenplay                   59547
BAFTA nominations                                    59547
Golden Globe Award for Best Screenplay               59547
Writers Guild Awards Winners & Nominees 2020-2013    59547
director                                             584

In [11]:
#Fehlende Werte in Prozent
(df.isna().sum()/len(df)).apply('{0:.4%}'.format)

name                                                  1.4040%
character                                            13.3554%
role                                                  1.4040%
title                                                 0.0000%
description                                           0.0364%
runtime                                               0.0000%
genres_x                                              0.0000%
production_countries_x                                0.0000%
imdb_score                                            0.0000%
imdb_votes                                            0.1174%
tmdb_score                                            0.0000%
Academy Awards adapted screenplay                    98.4736%
Academy Awards original screenplay                   98.4736%
BAFTA nominations                                    98.4736%
Golden Globe Award for Best Screenplay               98.4736%
Writers Guild Awards Winners & Nominees 2020-2013    98.4736%
director

In [12]:
#Wir haben immer noch sehr viel fehlende Werte, wenn wir alle bereinigen hat unser Datensatz keine Zeilen mehr

df_clean = df.dropna()

df_clean

Empty DataFrame
Columns: [name, character, role, title, description, runtime, genres_x, production_countries_x, imdb_score, imdb_votes, tmdb_score, Academy Awards adapted screenplay, Academy Awards original screenplay, BAFTA nominations, Golden Globe Award for Best Screenplay, Writers Guild Awards Winners & Nominees 2020-2013, director, cast, country, genres]
Index: []

In [13]:
#Daher müssen wir leider noch mehr Spalten entfernen

df = df.drop(['Academy Awards adapted screenplay', 'Academy Awards original screenplay', 'BAFTA nominations', 'Golden Globe Award for Best Screenplay', 'Writers Guild Awards Winners & Nominees 2020-2013', 'director', 'cast', 'country', 'genres'],1)

<ipython-input-13-565d01ae0318>:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop(['Academy Awards adapted screenplay', 'Academy Awards original screenplay', 'BAFTA nominations', 'Golden Globe Award for Best Screenplay', 'Writers Guild Awards Winners & Nominees 2020-2013', 'director', 'cast', 'country', 'genres'],1)


In [14]:
df.isna().sum()

name                       849
character                 8076
role                       849
title                        0
description                 22
runtime                      0
genres_x                     0
production_countries_x       0
imdb_score                   0
imdb_votes                  71
tmdb_score                   0
dtype: int64

In [15]:
df = df.dropna()

In [16]:
df.isna().sum()

name                      0
character                 0
role                      0
title                     0
description               0
runtime                   0
genres_x                  0
production_countries_x    0
imdb_score                0
imdb_votes                0
tmdb_score                0
dtype: int64

In [17]:
df

name                                          character  \
17545  Graham Chapman  King Arthur / Voice of God / Middle Head / Hic...   
17546     John Cleese  Second Swallow-Savvy Guard / The Black Knight ...   
17547       Eric Idle  Dead Collector / Peasant 1 / Sir Robin the Not...   
17548   Terry Gilliam  Patsy / Green Knight / Old Man from Scene 24 (...   
17549     Terry Jones  Dennis's Mother / Sir Bedevere / Left Head / P...   
...               ...                                                ...   
81465           Kalia                                              Kalia   
81466           Dholu                                              Dholu   
81467           Bholu                                              Bholu   
81468            Raju                                               Raju   
81476         Emicida                                               Self   

        role                              title  \
17545  ACTOR    Monty Python and the Holy Grail   
17546  ACTOR    Monty Python and the Holy Grail   
17547  ACTOR    Monty Python and the Holy Grail   
17548  ACTOR    Monty Python and the Holy Grail   
17549  ACTOR    Monty Python and the Holy Grail   
...      ...                                ...   
81465  ACTOR           Chhota Bheem Aur Hanuman   
81466  ACTOR           Chhota Bheem Aur Hanuman   
81467  ACTOR           Chhota Bheem Aur Hanuman   
81468  ACTOR           Chhota Bheem Aur Hanuman   
81476  ACTOR  Emicida AmarElo Live in Sao Paulo   

                                             description  runtime  \
17545  King Arthur, accompanied by his squire, recrui...     91.0   
17546  King Arthur, accompanied by his squire, recrui...     91.0   
17547  King Arthur, accompanied by his squire, recrui...     91.0   
17548  King Arthur, accompanied by his squire, recrui...     91.0   
17549  King Arthur, accompanied by his squire, recrui...     91.0   
...                                                  ...      ...   
81465  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81466  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81467  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81468  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81476  To the beat of his avant-garde and eclectic rh...     99.0   

                              genres_x production_countries_x  imdb_score  \
17545  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17546  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17547  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17548  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17549  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
...                                ...                    ...         ...   
81465         ['animation', 'fantasy']                 ['IN']         7.3   
81466         ['animation', 'fantasy']                 ['IN']         7.3   
81467         ['animation', 'fantasy']                 ['IN']         7.3   
81468         ['animation', 'fantasy']                 ['IN']         7.3   
81476       ['documentation', 'music']                 ['BR']         7.6   

       imdb_votes  tmdb_score  
17545    534486.0         7.8  
17546    534486.0         7.8  
17547    534486.0         7.8  
17548    534486.0         7.8  
17549    534486.0         7.8  
...           ...         ...  
81465        36.0         6.2  
81466        36.0         6.2  
81467        36.0         6.2  
81468        36.0         6.2  
81476        44.0         9.8  

[52323 rows x 11 columns]

In [18]:
#Nach dem wir die Nullwerte bereinigt haben wollen wir noch nach Duplikaten suchen

df.duplicated().sum()

3

In [19]:
#Da es nur 3 Zeilen sind können wir diese entfernen

df = df.drop_duplicates()

In [20]:
#Durch die Angabe der Schauspieler sind die Titel mehrmals enthalten, wir wollen sehen wie viele Filme in dem Datensatz enthalten sind:

uniques = pd.unique(df[['title']].values.ravel())

len(uniques)

3950

In [21]:
df

name                                          character  \
17545  Graham Chapman  King Arthur / Voice of God / Middle Head / Hic...   
17546     John Cleese  Second Swallow-Savvy Guard / The Black Knight ...   
17547       Eric Idle  Dead Collector / Peasant 1 / Sir Robin the Not...   
17548   Terry Gilliam  Patsy / Green Knight / Old Man from Scene 24 (...   
17549     Terry Jones  Dennis's Mother / Sir Bedevere / Left Head / P...   
...               ...                                                ...   
81465           Kalia                                              Kalia   
81466           Dholu                                              Dholu   
81467           Bholu                                              Bholu   
81468            Raju                                               Raju   
81476         Emicida                                               Self   

        role                              title  \
17545  ACTOR    Monty Python and the Holy Grail   
17546  ACTOR    Monty Python and the Holy Grail   
17547  ACTOR    Monty Python and the Holy Grail   
17548  ACTOR    Monty Python and the Holy Grail   
17549  ACTOR    Monty Python and the Holy Grail   
...      ...                                ...   
81465  ACTOR           Chhota Bheem Aur Hanuman   
81466  ACTOR           Chhota Bheem Aur Hanuman   
81467  ACTOR           Chhota Bheem Aur Hanuman   
81468  ACTOR           Chhota Bheem Aur Hanuman   
81476  ACTOR  Emicida AmarElo Live in Sao Paulo   

                                             description  runtime  \
17545  King Arthur, accompanied by his squire, recrui...     91.0   
17546  King Arthur, accompanied by his squire, recrui...     91.0   
17547  King Arthur, accompanied by his squire, recrui...     91.0   
17548  King Arthur, accompanied by his squire, recrui...     91.0   
17549  King Arthur, accompanied by his squire, recrui...     91.0   
...                                                  ...      ...   
81465  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81466  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81467  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81468  The grouchy Makhichoos, whoe hates Diwali, is ...     67.0   
81476  To the beat of his avant-garde and eclectic rh...     99.0   

                              genres_x production_countries_x  imdb_score  \
17545  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17546  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17547  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17548  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
17549  ['fantasy', 'action', 'comedy']                 ['GB']         8.2   
...                                ...                    ...         ...   
81465         ['animation', 'fantasy']                 ['IN']         7.3   
81466         ['animation', 'fantasy']                 ['IN']         7.3   
81467         ['animation', 'fantasy']                 ['IN']         7.3   
81468         ['animation', 'fantasy']                 ['IN']         7.3   
81476       ['documentation', 'music']                 ['BR']         7.6   

       imdb_votes  tmdb_score  
17545    534486.0         7.8  
17546    534486.0         7.8  
17547    534486.0         7.8  
17548    534486.0         7.8  
17549    534486.0         7.8  
...           ...         ...  
81465        36.0         6.2  
81466        36.0         6.2  
81467        36.0         6.2  
81468        36.0         6.2  
81476        44.0         9.8  

[52320 rows x 11 columns]

Wir erhalten einen Datensatz mit 3950 Filmen. Wenn wir diesen Datensatz mit den ursprünglichen Daten vergleichen sehen wir, dass er sehr ähnlich ist zu den Datensatz credits und titles und der Join der vielen Datensätze keinen Mehrwer. Daher würde wir im nächsten Schritt nur diese beiden Datensätze joinen um eine solide Datenbasis für unser Modell zu erhalten.

In [22]:
#Datei abspeichern:

df.to_csv("Df_Join_5_bereinigt.csv", index=False)